File: Sources/PLayout/Model/PlateFormat.swift
import Foundation

/// Physical shape of a microplate. Row/column counts follow ANSI/SLAS standards.
struct PlateFormat: Codable, Hashable, Identifiable {
    var rows: Int
    var cols: Int

    var id: String { "\(rows)x\(cols)" }
    var wellCount: Int { rows * cols }
    var name: String { "\(wellCount)-well" }
    var detailedName: String { "\(wellCount)-well  (\(rows)×\(cols))" }

    static let well6 = PlateFormat(rows: 2, cols: 3)
    static let well12 = PlateFormat(rows: 3, cols: 4)
    static let well24 = PlateFormat(rows: 4, cols: 6)
    static let well48 = PlateFormat(rows: 6, cols: 8)
    static let well96 = PlateFormat(rows: 8, cols: 12)
    static let well384 = PlateFormat(rows: 16, cols: 24)
    static let well1536 = PlateFormat(rows: 32, cols: 48)

    static let standard: [PlateFormat] = [
        .well6, .well12, .well24, .well48, .well96, .well384, .well1536,
    ]

    /// Bounds for user-defined plates. The ceiling keeps rendering responsive and
    /// stays far inside Excel's sheet limits on export.
    static let rowRange = 1...64
    static let columnRange = 1...96

    var isStandard: Bool { Self.standard.contains(self) }

    init(rows: Int, cols: Int) {
        self.rows = min(max(rows, Self.rowRange.lowerBound), Self.rowRange.upperBound)
        self.cols = min(max(cols, Self.columnRange.lowerBound), Self.columnRange.upperBound)
    }

    /// Clamped on the way in as well, so a hand-edited document can never produce a
    /// zero-sized plate that would divide by zero during layout.
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.init(
            rows: try container.decode(Int.self, forKey: .rows),
            cols: try container.decode(Int.self, forKey: .cols)
        )
    }

    func index(row: Int, col: Int) -> Int { row * cols + col }
    func row(of index: Int) -> Int { index / cols }
    func col(of index: Int) -> Int { index % cols }
    func contains(row: Int, col: Int) -> Bool {
        row >= 0 && row < rows && col >= 0 && col < cols
    }
}

enum WellNaming {
    /// 0 -> "A", 25 -> "Z", 26 -> "AA" (needed for 1536-well plates, which run A..AF).
    static func rowLabel(_ row: Int) -> String {
        guard row >= 0 else { return "?" }
        var out = ""
        var n = row
        repeat {
            out = String(UnicodeScalar(UInt8(65 + n % 26))) + out
            n = n / 26 - 1
        } while n >= 0
        return out
    }

    /// Inverse of `rowLabel`. Returns nil for anything that is not pure letters.
    static func rowIndex(_ label: String) -> Int? {
        let s = label.trimmingCharacters(in: .whitespaces).uppercased()
        guard !s.isEmpty, s.allSatisfy({ $0.isLetter && $0.isASCII }) else { return nil }
        var n = 0
        for ch in s.unicodeScalars {
            n = n * 26 + (Int(ch.value) - 64)
        }
        return n - 1
    }

    static func colLabel(_ col: Int, padded: Bool) -> String {
        padded ? String(format: "%02d", col + 1) : "\(col + 1)"
    }

    static func wellLabel(row: Int, col: Int, padded: Bool) -> String {
        rowLabel(row) + colLabel(col, padded: padded)
    }

    /// Parses "A1", "a01", "H12" into a position.
    static func parseWell(_ text: String) -> (row: Int, col: Int)? {
        let s = text.trimmingCharacters(in: .whitespaces)
        guard !s.isEmpty else { return nil }
        let letters = s.prefix { $0.isLetter }
        let digits = s.dropFirst(letters.count)
        guard !letters.isEmpty, !digits.isEmpty,
              let row = rowIndex(String(letters)),
              let num = Int(digits), num >= 1
        else { return nil }
        return (row, num - 1)
    }
}

File: python_port/playout/model/plate_format.py
"""PlateFormat and WellNaming — mirrors `Sources/PLayout/Model/PlateFormat.swift`.

Pure Python, no Qt. A `PlateFormat` is pure geometry (rows × cols); custom *names* live on
plate templates (`templates.py`), because every "did the format change" guard compares
dimensions alone.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

ROW_RANGE = (1, 64)
COLUMN_RANGE = (1, 96)


def _clamp(value: int, bounds: tuple[int, int]) -> int:
    lo, hi = bounds
    return min(max(int(value), lo), hi)


@dataclass(frozen=True, slots=True)
class PlateFormat:
    """Physical shape of a microplate. Clamped on construction *and* on decode, so a
    hand-edited document can never produce a zero-sized plate."""

    rows: int
    cols: int

    def __post_init__(self) -> None:
        object.__setattr__(self, "rows", _clamp(self.rows, ROW_RANGE))
        object.__setattr__(self, "cols", _clamp(self.cols, COLUMN_RANGE))

    # -- identity / names -------------------------------------------------------
    @property
    def id(self) -> str:
        return f"{self.rows}x{self.cols}"

    @property
    def well_count(self) -> int:
        return self.rows * self.cols

    @property
    def name(self) -> str:
        return f"{self.well_count}-well"

    @property
    def detailed_name(self) -> str:
        return f"{self.well_count}-well  ({self.rows}×{self.cols})"

    @property
    def is_standard(self) -> bool:
        return self in STANDARD

    @classmethod
    def standard_matching(cls, rows: int, cols: int) -> "PlateFormat | None":
        fmt = cls(rows, cols)
        return fmt if fmt in STANDARD else None

    # -- indexing (row-major everywhere, exports included) -----------------------
    def index(self, row: int, col: int) -> int:
        return row * self.cols + col

    def row_of(self, index: int) -> int:
        return index // self.cols

    def col_of(self, index: int) -> int:
        return index % self.cols

    def contains(self, row: int, col: int) -> bool:
        return 0 <= row < self.rows and 0 <= col < self.cols

    # -- codec -------------------------------------------------------------------
    def to_json(self) -> dict[str, Any]:
        return {"rows": self.rows, "cols": self.cols}

    @classmethod
    def from_json(cls, data: Any) -> "PlateFormat":
        """Both keys are required (a missing one is a decode error), values are clamped."""
        from playout.model.layout import LayoutDecodeError, require_int

        if not isinstance(data, dict):
            raise LayoutDecodeError("PlateFormat must be an object")
        return cls(require_int(data, "rows", "PlateFormat"), require_int(data, "cols", "PlateFormat"))


WELL6 = PlateFormat(2, 3)
WELL12 = PlateFormat(3, 4)
WELL24 = PlateFormat(4, 6)
WELL48 = PlateFormat(6, 8)
WELL96 = PlateFormat(8, 12)
WELL384 = PlateFormat(16, 24)
WELL1536 = PlateFormat(32, 48)

STANDARD: tuple[PlateFormat, ...] = (WELL6, WELL12, WELL24, WELL48, WELL96, WELL384, WELL1536)

# Class-level aliases so callers can write `PlateFormat.WELL96` / `PlateFormat.STANDARD`.
PlateFormat.WELL6 = WELL6  # type: ignore[attr-defined]
PlateFormat.WELL12 = WELL12  # type: ignore[attr-defined]
PlateFormat.WELL24 = WELL24  # type: ignore[attr-defined]
PlateFormat.WELL48 = WELL48  # type: ignore[attr-defined]
PlateFormat.WELL96 = WELL96  # type: ignore[attr-defined]
PlateFormat.WELL384 = WELL384  # type: ignore[attr-defined]
PlateFormat.WELL1536 = WELL1536  # type: ignore[attr-defined]
PlateFormat.STANDARD = STANDARD  # type: ignore[attr-defined]


def is_standard(rows: int, cols: int) -> bool:
    return PlateFormat(rows, cols) in STANDARD


PlateFormat.is_standard_shape = staticmethod(is_standard)  # type: ignore[attr-defined]


class WellNaming:
    """A1-style well names. Rows are bijective base-26 (A…Z, AA…), columns 1-based."""

    @staticmethod
    def row_label(row: int) -> str:
        """0 → "A", 25 → "Z", 26 → "AA" (a 1536-well plate runs A…AF)."""
        if row < 0:
            return "?"
        out = ""
        n = row
        while True:
            out = chr(65 + n % 26) + out
            n = n // 26 - 1
            if n < 0:
                break
        return out

    @staticmethod
    def row_index(label: str) -> int | None:
        """Inverse of `row_label`; None for anything that is not pure ASCII letters."""
        s = label.strip().upper()
        if not s or not all("A" <= ch <= "Z" for ch in s):
            return None
        n = 0
        for ch in s:
            n = n * 26 + (ord(ch) - 64)
        return n - 1

    @staticmethod
    def col_label(col: int, padded: bool) -> str:
        return f"{col + 1:02d}" if padded else str(col + 1)

    @staticmethod
    def well_label(row: int, col: int, padded: bool) -> str:
        return WellNaming.row_label(row) + WellNaming.col_label(col, padded)

    @staticmethod
    def parse_well(text: str) -> tuple[int, int] | None:
        """Parses "A1", "a01", " H12 " into (row, col); rejects letter-only and digit-only."""
        s = text.strip()
        if not s:
            return None
        i = 0
        while i < len(s) and s[i].isalpha():
            i += 1
        letters, digits = s[:i], s[i:]
        if not letters or not digits or not digits.lstrip("+").isdigit() or digits.count("+") > 1:
            return None
        row = WellNaming.row_index(letters)
        if row is None:
            return None
        num = int(digits)
        if num < 1:
            return None
        return (row, num - 1)
